### ISSUE TITLE
allow authz plugins to register routes they will listen on

### ISSUE THREAD

**Issue by runcom** [docker/docker#28244]
Basically, in a chain of i.e. 10 authz plugins, the daemon will call those out every requests. This makes CPU utilization of just one plugin being really high (if dockerd is at 20%, the plugin will be at just 5%).
This is the case even with an "empty" plugin which just reply "authorize: true".

Can we instead have plugins register routes they will listen on and just call them out from the daemon when the route matches?

**Comment by runcom** [comment]
ping @liron-l

**Comment by icecrime** [comment]
Cc @anusha-ragunathan @riyazdf.

**Comment by liron-l** [comment]
@runcom, I'm not sure I understand how will this improve performance. I assume any authentication plugin will require handling `create` and `start` APIs (otherwise no security model), and as far as I know, the only additional call kubernetes does when creating a container is the `docker version` API.
Assuming that any authentication plugin will require handling `create` and `start` APIs (otherwise no security model) this only saves one call.

Also, I think this has potential to cause some security issues. E.g., when customer upgrades to a newer docker version (which comes with a new set of APIs) but does not update the authorization plugin, then all new APIs will become exposed.

**Comment by runcom** [comment]
I see your point. The other plugin we tested was just meant to block "push" operations based on the image being pushed. The plugin was simply checking if the route hit was the "push" route API, otherwise return "authorized: true". No handling of create and start but yet, this caused a lot of cpu utilization by the plugin itself because the plugin was called on every api request (not just the push operation).

**Comment by anusha-ragunathan** [comment]
What level will routes be limited to? Only base endpoints (eg containers, images)? Or something more granular (eg. containers/create, image/push)? 
However this is achieved, we still have the issue @liron-l points out, which is that authz plugins dont have a strict correlation to the daemon API version. Eg. an authz plugin written for daemon API version 1.20 can run with docker 1.13 (API version 1.13) and allow all *volume* endpoints, thereby exposing newly added `volume prune`. 

IMO, before we innovate on having an access control list specification for authz plugins, we should resolve the API versioning and have some sort of a handshake protocol.

**Comment by riyazdf** [comment]
I agree that this has tight coupling with limiting the remote API versions for an authz plugin, linking the issue for that here: https://github.com/docker/docker/issues/27959

@runcom @liron-l what do you think of making all of the endpoints that the authz plugin doesn't listen to being default-deny?  It would make `docker version` a hassle but might be helpful for making the whitelist of allowed APIs for an authz plugin more explicit.

**Comment by runcom** [comment]
The plugin must a way to whitelist endpoints right? Is there anything like that already? I'm fine having a deny-all policy also. I expect the daemon not to call out the plugin if the endpoint is not whitelisted, correct?

**Comment by anusha-ragunathan** [comment]
Default deny of endpoints that the plugin doesnt register with, will cause issue with chaining.
